[bugfix][serve][llm] Fix port collisions for TP/PP with NIXL/LMCache

nrghosh opened the pull request:
<!-- Thank you for contributing to Ray! 🚀 -->
<!-- Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->
<!-- 💡 Tip: Mark as draft if you want early feedback, or ready for review when it's complete -->

## Description

Extends port collision fix to Tensor Parallelism (TP) and Pipeline Parallelism (PP) scenarios. Previous fix (PR #55802) only addressed Data Parallelism by using explicit `data_parallel_rank`.

**Changes:**
- `base.py`: Added `_compute_port_offset()` method with fallback logic
  * Priority 1: Use `data_parallel_rank` if set (DP case)
  * Priority 2: Use Ray Serve [`replica_rank`](https://docs.ray.io/en/master/serve/advanced-guides/replica-ranks.html) for deterministic offset (TP/PP case)
  * Fallback: Return 0
- `nixl_connector.py`: Use `_compute_port_offset()` instead of direct `dp_rank` access
- `lmcache_connector_v1.py`: Simplified to use string-based port naming with random suffix

**Fixes port collision errors in TP/PP deployments:**
- Multiple workers no longer bind to same port
- Prevents `NIXL_ERR_BACKEND` and ZMQ errors
- Unblocks deployment with `PP size > 1`

**Reproduction:**
Deployed Ray Serve with `pipeline_parallel_size=2` and NIXL on Ray 3.0.0.dev0 (8 x L4 GPU cluster). Before fix, all workers used identical port (e.g., 52910), causing `NIXL_ERR_BACKEND`. Logs showed:
  `Creating v1 connector with engine_id: ...-52910 [repeated 3x]`
After fix, each worker receives unique port via `replica_rank`, eliminating collisions.


## Related issues

Addresses https://github.com/ray-project/ray/issues/55775
Addresses https://github.com/vllm-project/vllm/issues/20980#

## Types of change

- [x] Bug fix 🐛
- [ ] New feature ✨
- [ ] Enhancement 🚀
- [ ] Code refactoring 🔧
- [ ] Documentation update 📖
- [ ] Chore 🧹
- [ ] Style 🎨

## Checklist

**Does this PR introduce breaking changes?**
- [ ] Yes ⚠️
- [x] No

**Testing:**
- [x] Added/updated tests for my changes
- [x] Tested the changes manually
- [ ] This PR is not tested ❌ _(please explain why)_

**Code Quality:**
- [x] Signed off every commit (`git commit -s`)
- [x] Ran pre-commit hooks ([setup guide](https://docs.ray.io/en/latest/ray-contribute/getting-involved.html#lint-and-formatting))

**Documentation:**
- [ ] Updated documentation (if applicable) ([contribution guide](https://docs.ray.io/en/latest/ray-contribute/docs.html))
- [ ] Added new APIs to `doc/source/` (if applicable)

## Additional context

### Code Changes

**NIXL Connector** - Before:
```python
dp_rank = self.llm_config.engine_kwargs.get("data_parallel_rank", 0)
port = base_port + dp_rank  # Always 0 for TP/PP!
```

**NIXL Connector** - After:
```python
port = base_port + self._compute_port_offset()  # Works for DP/TP/PP
```

**`_compute_port_offset()` Implementation:**
```python
def _compute_port_offset(self) -> int:
    # Priority 1: Use explicit DP rank when available
    dp_rank = self.llm_config.engine_kwargs.get("data_parallel_rank")
    if isinstance(dp_rank, int) and dp_rank >= 0:
        return dp_rank
    
    # Priority 2: Fall back to Serve replica rank for TP/PP cases
    try:
        rc = serve.get_replica_context()
        if rc and hasattr(rc, "rank"):
            return rc.rank
    except Exception:
        pass
    
    return 0
```

**LMCache Connector** - Simplified approach:
```python
# Always use string-based naming with random suffix for uniqueness
lmcache_rpc_port_value = str(base_value) + self._get_unique_suffix()
```

### Backward Compatibility

- DP deployments continue using explicit `data_parallel_rank` (priority 1)
- TP/PP deployments now use `replica_rank` from Ray Serve (priority 2)
- LMCache uses string-based port naming with random suffix (unchanged behavior)
- Zero fallback maintains current behavior when neither rank is available

nrghosh committed:
serve.llm: Fix port collisions for TP/PP with NIXL/LMCache

Extends port collision fix to Tensor Parallelism (TP) and Pipeline
Parallelism (PP) scenarios. Previous fix (PR #55802) only addressed
Data Parallelism by using explicit data_parallel_rank.

Changes:
- base.py: Added _compute_port_offset() method with fallback logic
  * Priority 1: Use data_parallel_rank if set (DP case)
  * Priority 2: Hash replica_tag for deterministic offset (TP/PP case)
  * Fallback: Return 0
- nixl_connector.py: Use _compute_port_offset() instead of dp_rank
- lmcache_connector_v1.py: Add numeric port support with offset logic

Fixes port collision errors in TP/PP deployments:
- Multiple workers no longer bind to same port
- Prevents NIXL_ERR_BACKEND and ZMQ errors
- Enables successful deployment with pipeline_parallel_size > 1

Reproduction:
Deployed Ray Serve with pipeline_parallel_size=2 and NIXL on Ray
3.0.0.dev0 (8 x L4 GPU cluster). Before fix, all workers used identical
port (e.g., 52910), causing NIXL_ERR_BACKEND. Logs showed:
  'Creating v1 connector with engine_id: ...-52910 [repeated 3x]'
After fix, each worker receives unique port via replica tag hashing,
eliminating collisions.

Related: #55775
Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh reviewed (commented):
Failing `llm_serve_vllm_integration_tests::test_deepseek_model `  release test with 

```
KeyError: Deployment(name='LLMServer:deepseek-ai--DeepSeek-V2-Lite', app='default')
```

But passes locally on feature branch

```
PASSED

============================================================================ 1 passed in 57.16s ============================================================================
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225270) WARNING 10-16 17:45:05 [fused_moe.py:798] Using default MoE config. Performance might be sub-optimal! Config file not found at ['/home/ray/anaconda3/lib/python3.11/site-packages/vllm/model_executor/layers/fused_moe/configs/E=64,N=704,device_name=NVIDIA_L4.json'] [repeated 4x across cluster]
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) WARNING 10-16 17:44:59 [symm_mem.py:58] SymmMemCommunicator: Device capability 8.9 not supported, communicator is not available.
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) INFO 10-16 17:44:59 [custom_all_reduce.py:35] Skipping P2P check and trusting the driver's P2P report.
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) WARNING 10-16 17:44:59 [custom_all_reduce.py:154] Custom allreduce is disabled because your platform lacks GPU P2P capability or P2P test failed. To silence this warning, specify disable_custom_all_reduce=True explicitly.
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) INFO 10-16 17:44:59 [parallel_state.py:1208] rank 1 in world size 4 is assigned as DP rank 0, PP rank 0, TP rank 1, EP rank 1
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) WARNING 10-16 17:44:59 [topk_topp_sampler.py:66] FlashInfer is not available. Falling back to the PyTorch-native implementation of top-p & top-k sampling. For the best performance, please install FlashInfer.
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) INFO 10-16 17:44:59 [gpu_model_runner.py:2602] Starting to load model deepseek-ai/DeepSeek-V2-Lite...
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) INFO 10-16 17:44:59 [gpu_model_runner.py:2634] Loading model from scratch...
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) INFO 10-16 17:44:59 [utils.py:125] Hidden layers were unevenly partitioned: [14,13]. This can be manually overridden using the VLLM_PP_LAYER_PARTITION environment variable
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) INFO 10-16 17:44:59 [cuda.py:297] Using Triton MLA backend on V1 engine.
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) INFO 10-16 17:45:00 [weight_utils.py:392] Using model weights format ['*.safetensors']
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225271) INFO 10-16 17:45:06 [gpu_worker.py:298] Available KV cache memory: 12.54 GiB [repeated 4x across cluster]
(ServeReplica:default:LLMServer:deepseek-ai--DeepSeek-V2-Lite pid=224669) (EngineCore_DP0 pid=225159) (RayWorkerWrapper pid=225270) WARNING 10-16 17:45:07 [cudagraph_dispatcher.py:106] cudagraph dispatching keys are not initialized. No cudagraph will be used. [repeated 2x across cluster]
(base) ray@ip-10-0-167-142:~/default/work/ray$ 
(base) ray@ip-10-0-167-142:~/default/work/ray$ git status
On branch nrghosh/pp-tp-kv-port-offset
nothing to commit, working tree clean
```

nrghosh committed:
Merge branch 'master' into nrghosh/pp-tp-kv-port-offset

kouroshHakha commented on python/ray/llm/_internal/serve/deployments/llm/vllm/kv_transfer_backends/base.py:
```diff
@@ -35,6 +35,38 @@ def _get_unique_suffix(self, len: int = 6) -> str:
+            from ray import serve  # type: ignore
+
+            rc = serve.get_replica_context()
+            if rc and getattr(rc, "replica_tag", None):
```
what is replica_tag? maybe it's better to use the new replica_rank API introduced recently https://docs.ray.io/en/master/serve/advanced-guides/replica-ranks.html. This also would allow use to unify priority 1 and 2. in the dp case, replica_rank would correspond to the dp_rank or at least some shuffled order of that?

nrghosh replied:
switched to using `rc.rank` from the replica context API - much cleaner

for DP deployments we'll still use `data_parallel_rank`since it's explicitly set by DPServer, but for TP/PP cases, `replica_rank` is a stable offset

kouroshHakha commented on python/ray/llm/_internal/serve/deployments/llm/vllm/kv_transfer_backends/base.py:
```diff
@@ -35,6 +35,38 @@ def _get_unique_suffix(self, len: int = 6) -> str:
+
+            rc = serve.get_replica_context()
+            if rc and getattr(rc, "replica_tag", None):
+                import zlib
```
is this an extra dependency? We should import these things at top so that they raise errors on ci/cd tests.

kouroshHakha commented on python/ray/llm/_internal/serve/deployments/llm/vllm/kv_transfer_backends/base.py:
```diff
@@ -35,6 +35,38 @@ def _get_unique_suffix(self, len: int = 6) -> str:
         """
         return "".join(random.choices(string.ascii_letters + string.digits, k=len))
 
+    def _compute_port_offset(self) -> int:
```
we should just use the replica rank to do this I feel like.

nrghosh replied:
yep

now `_compute_port_offset()` uses `replica_rank` from replica context instead of hashing approach 

so now logic is:
1. Use `data_parallel_rank` if explicitly set (DP deployments via DPServer)
2. Fall back to `replica_rank` from serve context (TP/PP deployments)
3. Return 0 as final fallback

kouroshHakha commented on python/ray/llm/_internal/serve/deployments/llm/vllm/kv_transfer_backends/lmcache_connector_v1.py:
```diff
@@ -41,21 +43,34 @@ def setup(self) -> None:
+        # Determine the desired style of RPC port configuration.
+        # If user passes a numeric base (e.g., 50000), add a deterministic
+        # rank-based offset to avoid collisions across DP/TP/PP.
+        # Otherwise, default to string-based name + random suffix.
```
why would there be different types for LMCACHE_RPC_PORT_FIELD_NAME? If it is a name, it should always be considered as string?

nrghosh replied:
- simplified to always treat the port as a string and append random suffix
- int-based was mirroring how NIXL does port offsetting, string-based (LMCache) works fine with random suffix

kouroshHakha commented on python/ray/llm/_internal/serve/deployments/llm/vllm/kv_transfer_backends/lmcache_connector_v1.py:
```diff
@@ -41,21 +43,34 @@ def setup(self) -> None:
         kv_connector_extra_config[
             LMCacheConnectorV1Backend.LMCACHE_RPC_PORT_FIELD_NAME
-        ] = lmcache_rpc_port
+        ] = lmcache_rpc_port_value
```
based on this you just need a unique name for each replica? regardless of whether the port is passed in as int or string? 

Just think of some way to trim this complexity down. I feel like these if/else conditions are more confusing than making things simpler.

nrghosh replied:
Simplified, yes - LMCacheConnector now always treats the base value as a string + appends random suffix

nrghosh committed:
wip -  use replica_rank and simplify LMCache

- Use replica_rank API instead of hashing approach
- Simplify LMCache connector by just keeping string approach
- Update comments / lint

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge remote-tracking branch 'upstream/master' into nrghosh/pp-tp-kv-port-offset

kouroshHakha reviewed (approved):
Better. Thanks
